File: gems/espalier/test/fixtures/big_o/recursive_suffix_rescan.rb
class RecursiveSuffixRescan
  def initialize(items)
    @items = items
    @position = 0
  end

  def walk
    return if @position >= @items.length

    scan_remaining
    @position += 1
    walk
  end

  def scan_remaining
    cursor = @position
    while cursor < @items.length
      consume(@items[cursor])
      cursor += 1
    end
  end
end

File: gems/espalier/test/fixtures/big_o/recursive_suffix_rescan.py
class PyRecursiveSuffixRescan:
    def __init__(self, items):
        self.items = items
        self.position = 0

    def walk(self):
        if self.position >= len(self.items):
            return

        self.scan_remaining()
        self.position += 1
        self.walk()

    def scan_remaining(self):
        cursor = self.position
        while cursor < len(self.items):
            consume(self.items[cursor])
            cursor += 1
